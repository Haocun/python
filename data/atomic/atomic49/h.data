# This is a data file which can be used to input atomic data
# The first word defines the type of record.  The data which follows
# Element and Ion refers to that element and ion until the next element
# or ion is begun.  Continuation lines may be indicated by * in the first
# word in the line. 

# Lines beginning with # will be regarded as comments.

# If a line is blank it will be ignored.
# The order of the inputs can sometimes be important. In particular, A
# should be the first line for any element and Ion should be the first line
# for any ion
# 

# For elements, the format is
#Element  z   name  log abun (by number)
Element 1	H	12.00
#Element	2	He	10.93
#Element	6	C	8.52
#Element	7	N	7.96
#Element	8	0	8.82

#Element	10	Ne	7.92
#Element 11  Na  6.25
#Element 12  Mg  7.42
#Element	13	Al	6.39
#Element	14	Si	7.52
#Element	15	P	5.52
#Element	16	S	7.20
#Element 18  Ar  6.8
#Element 20  Ca  6.30
#Element	26	Fe	7.60
#Element 28  Ni  6.30

#For ions, the format is
# Ion  Element   z ionization State  g IP(in eV)
# However, the element name is not saved, indexing is based on z and istate
# Levels if they exist have the format
# Level Principle q #  g  excitation energy

Ion	H	1	1	2	13.598
Level 1  2        0
Level 2  8 10.19085
Level 3 18 12.07804
Level 4 32 12.7385625
Level 5 50 13.044288

Ion	H	1	2	2	1.e20

Ion	He	2	1	1	24.587
Level 1 1 0.0
Level 2 3 19.819
Level 2 1 20.615
Level 2 5 20.964
Level 2 3 20.964
Level 2 1 20.964
Level 2 3 21.218

Ion	He	2	2	2	54.416
Level 1  2 0
Level 2  8 40.7922
Level 3 18 48.31216
Level 4 32 50.905425
Level 5 50 52.177152

Ion	He	2	3	3	1.e20

Ion	C	6	1	3	11.26
Level 2 1 0
Level 2 3 2.0333605e-03
Level 2 5 5.3933649e-03
Level 2 5 1.26387
Level 2 1 2.684086
Level 2 5 4.182672

Ion	C	6	2	2	24.382
Level 2 2 0
Level 2 4 7.9350658e-03
Level 2 2 5.331397
Level 2 4 5.334075
Level 2 6 5.337658

Ion	C	6	3	1	47.883
Level 2 1 0
Level 2 1 6.486296
Level 2 3 6.489148
Level 2 5 6.496191

Ion	C	6	4	2	64.492
Level 2 2 0
Level 2 2 7.995100
Level 2 4 8.008378

Ion	C	6	5	1	392.08
Level 1 1 0
Ion	C	6	6	2	489.98
Level 1 2 0.0


Ion	C	6	7	2	1.e20

Ion	N	7	1	4	14.534
Ion	N	7	2	3	29.601
Ion	N	7	3	2	47.448
Ion	N	7	4	1	77.472
Ion	N	7	5	2	97.89
Ion	N	7	6	1	552.06
Ion	N	7	7	2	667.03
Ion	N	7	8	2	1.e20

Ion	O	8	1	3	13.618
Ion	O	8	2	4	35.117
Ion	O	8	3	3	54.900
Ion	O	8	4	2	77.412
Ion	O	8	5	1	113.896
Ion	O	8	6	2	138.115
Ion	O	8	7	1	739.30
Ion	O	8	8	2	871.39
Ion	O	8	9	2	1.e20

Ion Ne  10  1   1   21.564
Ion Ne  10  2   6   40.962
Ion Ne  10  3   9   63.45 
Ion Ne  10  4   4   97.11  
Ion Ne  10  5   9   126.21 
Ion Ne  10  6   6   157.93
Ion Ne  10  7   1   207.26
Ion Ne  10  8   2   239.09
Ion Ne  10  9   1   1195.8
Ion Ne  10  10  2   1362.2
Ion Ne  10  11  1   1.e20 

Ion Na  11  1    2  5.139
Ion Na  11  2    1  47.286
Ion Na  11  3    6  71.64
Ion Na  11  4    9  98.91
Ion Na  11  5    4  138.40
Ion Na  11  6    9  172.15
Ion Na  11  7    6  208.48
Ion Na  11  8    1  264.19
Ion Na  11  9    2  299.9
Ion Na  11  10   1  1.e20

Ion Mg  12  1    1  7.646
Ion Mg  12  2    2  15.035
Ion Mg  12  3    1  80.143
Ion Mg  12  4    6  109.31
Ion Mg  12  5    9  141.27
Ion Mg  12  6    4  186.51
Ion Mg  12  7    9  224.95
Ion Mg  12  8    6  265.92
Ion Mg  12  9    1  328.0
Ion Mg  12  10   2  367.5
Ion Mg  12  11   1  1.e20

Ion Al  13  1    6  5.986
Ion Al  13  2    1  18.826
Ion Al  13  3    2  28.448
Ion Al  13  4    1  119.99
Ion Al  13  5    6  153.75
Ion Al  13  6    9  190.47
Ion Al  13  7    4  241.44
Ion Al  13  8    9  284.59
Ion Al  13  9    6  330.2
Ion Al  13  10   1  398.6
Ion Al  13  11   2  442.0
Ion Al  13  12   1  1.e20



Ion	Si	14	1	3	8.151
Ion	Si	14	2	2	16.346
Ion	Si	14	3	1	33.492
Ion	Si	14	4	2	45.141
Ion	Si	14	5	1	166.77
Ion	Si	14	6	2	205.16
Ion	Si	14	7	3	246.5
Ion	Si	14	8	4	303.15
Ion	Si	14	9	3	351.05
Ion	Si	14	10	2	401.38
Ion	Si	14	11	1	476.15
Ion	Si	14	12	2	523.38
Ion	Si	14	13	2	2673.1
Ion	Si	14	14	2	1.e20

Ion	S	16	1	3	10.360
Ion	S	16	2	4	23.411
Ion	S	16	3	3	35.06
Ion	S	16	4	2	47.304
Ion	S	16	5	1	72.49
Ion	S	16	6	2	88.051
Ion	S	16	7	1	281.06
Ion	S	16	8	2	328.89
Ion	S	16	9	3	379.05
Ion	S	16	10	4	447.1
Ion	S	16	11	2	3494.10
Ion	S	16	12	2	1.e20

Ion Ar  18  1   1   15.759
Ion Ar  18  2   6   27.629
Ion Ar  18  3   9   40.74
Ion Ar  18  4   4   59.81
Ion Ar  18  5   9   75.04
Ion Ar  18  6   6   91.01
Ion Ar  18  7   1   124.4
Ion Ar  18  8   2   143.45
Ion Ar  18  9   1   422.6
Ion Ar  18  10  6   1.e20


Ion Ca  20  1   1   6.113
Ion Ca  20  2   2   11.871
Ion Ca  20  3   1   50.91
Ion Ca  20  4   6   67.15
Ion Ca  20  5   9   84.43
Ion Ca  20  6   4   108.78
Ion Ca  20  7   9   127.7
Ion Ca  20  8   6   147.4
Ion Ca  20  9   1   188.7
Ion Ca  20  10  2   211.3
Ion Ca  20  11  1   1.e20


# Ground state multiplicities are not all correct for Iron or ni
Ion Fe  26 1   25   7.870
Ion Fe  26 2   30   16.16
Ion Fe  26 3   25   30.651
Ion Fe  26 4   25   54.8 
Ion Fe  26 5   25   75.5
Ion Fe  26 6   25   100.0
Ion Fe  26 7   25   128.3
Ion Fe  26 8   25   151.12
Ion Fe  26 9   25   235.0
Ion Fe  26 10  25   262.1
Ion Fe  26 11  25   290.4
Ion Fe  26 12  25   331.0
Ion Fe  26 13  25   361.0
Ion Fe  26 14  25   392.0
Ion Fe  26 15  25   457.0
Ion Fe  26 16  25   459.0
Ion Fe  26 17  25   1.e20

Ion Ni  28 1   21   7.635
Ion Ni  28 2   21   18.168
Ion Ni  28 3   21   35.17
Ion Ni  28 4   21   54.9
Ion Ni  28 5   21   75.5
Ion Ni  28 6   21   108.
Ion Ni  28 7   21   134.
Ion Ni  28 8   21   164.
Ion Ni  28 9   21   193.
Ion Ni  28 10  21   224.6
Ion Ni  28 11  21   321.
Ion Ni  28 12  21   352.
Ion Ni  28 13  21   384.
Ion Ni  28 14  21   430.
Ion Ni  28 15  21   464.
Ion Ni  28 16  21   1.e20

# THIS NEXT SECTION IS REPLACED BY DATA IN PHOTO.DATA
#Photoionization crossections must be read in after all elements and ions
#have been defined!  Otherwise they will be discarded
#Format:  Photo  Element(z) Ion(istate) sigma_threshold beta alpha
#                   z          istate     10**-18 cm**2
#   (See Osterbrock)
# Phot 1 1  6.30 1.34 2.99
# Phot 2 1  7.83 1.66 2.05
# Phot 2 2  1.58 1.34 2.99
# Phot 6 1 12.2  3.32 2.0
# Phot 6 2  4.60 1.95 3.0
# Phot 6 3  1.60 2.6  3.0
# Phot 6 4  0.68 1.0  2.0
# Phot 7 1 11.4  4.29 2.0
# Phot 7 2  6.65 2.86 3.0
# Phot 7 3  2.06 1.63 3.0
# Phot 7 4  1.08 2.6  3.0
# Phot 7 5  0.48 1.0  2.0
# Next line a fudge, 3 lines in Osterbrock table combined
# Phot 8 1  9.06 4.31 1.5
# Phot 8 2  7.32 3.84 2.5
# Phot 8 3  3.65 2.01 3.0
# Phot 8 4  1.27 0.83 3.0
# Phot 8 5  0.78 2.6  3.0
# Phot 8 6  0.36 1.0  2.1


# Resonance line transitions
# Format:
# Line Element(z)  Ion(istate)  Wavelength  f
# Note that element and istate are used to determine the ion number in the
# structure.
#   (Resonance lines in a separate file)
