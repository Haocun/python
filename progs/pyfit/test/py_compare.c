


/**************************************************************************
                        Space Telescope Science Institute


  Synopsis:  py_compare 

  Description:	

  	This routine compare two python models with one another 

	Usage 

	py_compare [-m firstmodel nspec -m secondmodel nspec] pffile  

	Old

  	This is a routine to fit a spectrum to a series of python models.
	The intent is to fit a series of regions (lines) rather than
	the ovrall spectral shape.  An overall spectral shape is also
	fit from the sum of all the accepted regions, but this does not contribute to 
	the "best spectrum".  pyfit3 treats each inclination angle in the python 
	output files separately; it does not interpolate to produce a best fit for
	and intermediate inclination.

  Arguments:  (Input via .pf file)		

  The parameter file looks something like this 

first_model                             sscyg_fuse4    !spectrum to fit
first_spectrum
second_model
second_spectrum
number.of.independent.regions                    4     !number of regions to fit
wavelengths.to.fit                         920-960     
chi2.renorm                                       1    !Weight to give this region
wavelengths.to.fit                        960-1000
chi2.renorm                                       1
wavelengths.to.fit             1000-1025,1026-1050
chi2.renorm                                       1
wavelengths.to.fit                       1140-1185
chi2.renorm                                       1
output_file_root                             sscyg_fuse



  Returns:

The beginning of the output file (.results) looks something like this:

#  ratio  ave_chinu    ratio_ind...                      ratio_comb ave_norm angle params                                                                                          filename
   1.500  1.29e+03      2.089    1.983    3.411    1.754    1.945  1.15e+00  15.0  5.00e-10  5.00e-11  4.00e+00  0.00e+00  1.00e+00
 1.00e+00  1.00e+10 sscyg_kgrid/sscyg_00_00_00_00_00_00_00.spec


The program also produces the spectra in the file output_file_root.out.  This contaians the 
best overall fit as well as the individual spectra.

# sscyg_kgrid/sscyg_04_03_01_02_00_02_02.spec 30.0  3.439e+03
# sscyg_kgrid/sscyg_04_02_05_00_02_00_02.spec 75.0  7.062e+02 920-960
# sscyg_kgrid/sscyg_02_01_03_00_01_00_02.spec 75.0  4.465e+02 960-1000
# sscyg_kgrid/sscyg_04_03_01_02_00_02_02.spec 30.0  7.475e+02 1000-1025,1026-1050
# sscyg_kgrid/sscyg_04_04_00_01_02_02_00.spec 60.0  1.413e+02 1140-1185
# sscyg_kgrid/sscyg_01_00_01_01_00_01_00.spec 15.0  9.020e+02 920-960,960-1000,1000-1025,1026-1050,1140-1185
   904.500  2.433e-12  1.569e-11  2.379e-11  0    1.198e-11  0  1.797e-11  0  2.537e-11  0  1.477e-11  0  1.335e-11  0
   904.600  1.222e-12  6.032e-12  2.144e-11  0    1.108e-11  0  1.749e-11  0  2.287e-11  0  1.378e-11  0  1.354e-11  0
   904.700  7.269e-13  4.312e-12  2.085e-11  0    1.156e-11  0  1.735e-11  0  2.224e-11  0  1.462e-11  0  1.394e-11  0

The comment lines describe the various fits, and then each fit (and the region fitted) correspond to the 
various spectra

  Notes:


pyfit3 simply calculates the best normilization in each region and then
the chi**2 for each spectrum.  In terms of mattchin the model spectra to
the data, it uses the standard kslfit approach of binning the model spectra
into the binning of the data spectra.  

There is no allowance for reddening or anything else, although this would
not in principle be difficult to add.

It is not entirely obvious that pure chi**2 is the best apporach, but that
is the one used currently.

!! It would be desirable to print out the parameters of the various models
in the results.out file, but this would require storing this information
as the fits improved or finding the models in the list.  

  History:
090408	ksl	Adapted from pyfit3

 ************************************************************************/

#include <stdio.h>
#include <strings.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>


#include "log.h"



#define LINELENGTH 200
#define NCOMMENTS   10
#define NWAVES      14000
#define MAX_MOD    10
#define BIG        1.e32
#define MAXTOT     100000
#define NREG	10
#define MAXVAR	10



char waverange[NREG][LINELENGTH];

// frac_weight is reg_weight renormalized to 1
float reg_weight[NREG],
  frac_weight[NREG];

float overall, results[NREG];


#include "py_compare.h"
#include "templates.h"

int
main (argc, argv)
     int argc;
     char *argv[];
{

  FILE *optr, *fopen ();
  char parfile[LINELENGTH];
  char diagfile[LINELENGTH];

  char first_model[LINELENGTH];
  char second_model[LINELENGTH];
  int first_spec, second_spec;
  struct Spec first_spectrum, second_spectrum;

  char outname[LINELENGTH];
  char outputfile[LINELENGTH];
  char output_specfile[LINELENGTH];
  int i;
  int n, nreg;
  float x;


/* Initialize variables that will be read by rdpar.  This needs to be
 * some of these. Note that some of the names are important, because
 * they are used to determine whether the user has provided entries on
 * the command line*/

  for (n = 0; n < NREG; n++)
    {
      strcpy (waverange[n], "1525-1575");
      reg_weight[n] = 1.0;
    }
  strcpy (outname, "py_compare");


  strcpy (first_model, "Unknown");
  strcpy (second_model, "Unknown");
  first_spec = 0;
  second_spec = 0;



/* Parse the command line */
  if (argc == 1)
    {
      strcpy (parfile, "py_compare.pf");
    }
  else
    {
      for (i = 1; i < argc; i++)
	{
	  if (strcmp (argv[i], "-h") == 0)
	    {
	      printf ("Usage: py_compare [file.pf]\n");
	      exit (0);
	    }
	  else if (strcmp (argv[i], "-m1") == 0)
	    {
	      if (sscanf (argv[i + 1], "%s", first_model) != 1)
		{
		  Error
		    ("py_compare: Expected first model after -m switch\n");
		  exit (0);
		}
	      i++;
	      if (sscanf (argv[i + 1], "%d", &first_spec) != 1)
		{
		  Error
		    ("py_compare: Expected first spectrum after modelname\n");
		  exit (0);
		}
	      i++;

	    }
	  else if (strcmp (argv[i], "-m2") == 0)
	    {
	      if (sscanf (argv[i + 1], "%s", second_model) != 1)
		{
		  Error
		    ("py_compare: Expected first model after -m switch\n");
		  exit (0);
		}
	      i++;
	      if (sscanf (argv[i + 1], "%d", &second_spec) != 1)
		{
		  Error
		    ("py_compare: Expected first spectrum after modelname\n");
		  exit (0);
		}
	      i++;

	    }
	  else if (strcmp (argv[i], "-o") == 0)
	    {
	      if (sscanf (argv[i + 1], "%s", outname) != 1)
		{
		  Error
		    ("py_compare: Expected root for output files  after -o switch\n");
		  exit (0);
		}
	      i++;

	    }
	  else if (strncmp (argv[i], "-", 1) == 0)
	    {
	      Error ("python: Unknown switch %s\n", argv[i]);
	      exit (0);
	    }

	}


      /* The last argument is always the pf file unless there were no arguments */

      strcpy (parfile, argv[argc - 1]);
      if ((strstr (parfile, ".pf")) == NULL)
	strcat (parfile, ".pf");

    }

  /* Construct an outputfile name if it is not given but model list or source comes from command line */



/* Open a log file */

  get_root (diagfile, parfile);
  strcat (diagfile, ".log");
  Log_init (diagfile);

/* Get data from a parameter file */


  opar (parfile);


  if (strcmp (first_model, "Unknown") == 0)
    {
      rdstr ("First_model", first_model);
      rdint ("First_spectrum", &first_spec);
    }
  if (strcmp (second_model, "Unknown") == 0)
    {
      rdstr ("Second_model", second_model);
      rdint ("Second_spectrum", &second_spec);
    }

  rdint ("number.of.independent.regions", &nreg);
  for (n = 0; n < nreg; n++)
    {
      rdstr ("wavelengths.to.fit", waverange[n]);
      rdflo ("chi2.renorm", &reg_weight[n]);
    }

  if (strcmp (outname, "py_compare") == 0)
    {
      get_root (outname, parfile);

    }

  strcpy (outputfile, outname);
  strcat (outputfile, ".out");
  strcpy (output_specfile, outname);
  strcat (output_specfile, ".spec");

  cpar (parfile);

/* End of input section */

  /* Compute the fractional weights */
  x=0.0;
  for (n=0;n<nreg;n++){
	  x+=reg_weight[n];
  }
  for (n=0;n<nreg;n++){
	  frac_weight[n]=reg_weight[n]/x;
  }

  /* Read the models */

  if (!get_python_spec (first_model, first_spec, &first_spectrum))
    {
      Error ("Could not read %s spectrum %d \n", first_model, first_spectrum);
      exit (0);
    }


  if (!get_python_spec (second_model, second_spec, &second_spectrum))
    {
      Error ("Could not read %s spectrum %d \n", second_model,
	     second_spectrum);
      exit (0);
    }

  overall = 0;

  for (n = 0; n < nreg; n++)
    {
      parse_waverange (&first_spectrum, waverange[n]);

      /* Now do the fit */

      x = metric (&first_spectrum, &second_spectrum);

      /* Now report the results somehow */

      results[n] = x;

      overall += frac_weight[n] * x;



    }


  /* Report the results */

  optr=fopen(outputfile,"a");


  printf ("%20s %2d  %20s %2d %8.3f", first_model, first_spec, second_model,
	  second_spec, overall);
  for (n = 0; n < nreg; n++)
    {
      printf (" %8.3f ", results[n]);
    }

  printf ("\n");



  fprintf (optr,"%20s %2d  %20s %2d %8.3f", first_model, first_spec, second_model,
	  second_spec, overall);
  for (n = 0; n < nreg; n++)
    {
      fprintf (optr," %8.3f ", results[n]);
    }

  fprintf (optr,"\n");


  fclose(optr);


  exit (0);

}





/*
 * This version adapted to change what is passed to
 * the routine 
 */

int
parse_waverange (spec, waverange)
     SpecPtr spec;
     char waverange[];
{
  int i, n;
  int nintervals;
  float www, w1[10], w2[10];
  char *xxx;

/* parse the string */
  xxx = waverange;
  nintervals = 0;
  while (xxx != NULL && nintervals < 10)
    {
      sscanf (xxx, "%f-%f", &w1[nintervals], &w2[nintervals]);
      nintervals++;
      if ((xxx = index (xxx, ',')) == NULL)
	break;
      xxx++;
    }


  for (n = 0; n < spec->nw; n++)
    spec->good[n] = 0;

  spec->ng = 0;
  for (n = 0; n < spec->nw; n++)
    {
      www = spec->w[n];
      for (i = 0; i < nintervals; i++)
	{
	  // Next line modified to allow overlapping waveranges
	  if (w1[i] <= www && www <= w2[i] && spec->good[n] == 0)
	    {
	      spec->good[n] = 1;
	      spec->ng++;
	    }
	}
    }

  if (spec->ng < 1)
    {
      Error
	("parse_wavelength_range:  No data points in waverange %s\n",
	 waverange);
      exit (0);
    }

  return (0);

}


float
metric (one, two)
     SpecPtr one, two;
{
  int n;
  float x, y;

  x = 0;
  y = 0;

  for (n = 0; n < one->nw; n++)
    {
      if (one->good[n])
	{
	  x += fabs (one->f[n] - two->f[n]);
	  y += 0.5 * (one->f[n] + two->f[n]);
	}
    }

  if (y > 0.0)
    {
      return (x / y);
    }
  return (1.0);

}
