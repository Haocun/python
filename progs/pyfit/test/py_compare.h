

#define LINELENGTH 200
#define NCOMMENTS   10
#define NWAVES      14000
#define MAX_MOD    10
#define BIG        1.e32
#define MAXTOT     100000
#define NREG	10
#define MAXVAR	10

#include <stdio.h>
#include <string.h>

#define MAXWORD 80



typedef struct Spec
{
  char name[LINELENGTH];
  float ang;
  float w[NWAVES];
  float f[NWAVES];
  int good[NWAVES];
  int nw, ng;
} spec_dummy, *SpecPtr;


