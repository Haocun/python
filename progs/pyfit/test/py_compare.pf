First_model                             polar.spec
First_spectrum                                   1
Second_model                            polar.spec
Second_spectrum                                  2
number.of.independent.regions                    2
wavelengths.to.fit                        1220-1250
chi2.renorm                                       1
wavelengths.to.fit                        1525-1575
chi2.renorm                                       2
