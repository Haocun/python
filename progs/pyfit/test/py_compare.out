polar.spec 1  polar.spec 2 0.089619 0.100621  0.084118 
polar.spec 1  polar.spec 2 0.089619 0.100621  0.084118 
          polar.spec  1            polar.spec  2    0.090    0.101     0.084 
          polar.spec  1            polar.spec  2    0.090    0.101     0.084 
          polar.spec  2            polar.spec  3    0.146    0.168     0.136 
